### ISSUE TITLE
Similar to the attention of SKNet.

### ISSUE THREAD

**Issue by GuideWsp** [zhanghang1989/ResNeSt#4]
It seems that the Split-Attention Network is very similar to the attention of SKNet.

**Comment by zhanghang1989** [comment]
This work is inspired by pioneering work, and generalizes SENet/SKNet into featuremap groups and outperforms prior work.

![image](https://user-images.githubusercontent.com/8041160/78451538-03601d00-763b-11ea-97c9-bab10cc87e2c.png)
![image](https://user-images.githubusercontent.com/8041160/78451554-1246cf80-763b-11ea-8be0-e8c1ed84a637.png)

**Comment by zhanghang1989** [comment]
We refer to the number of splits as Radix. Here is Radix-major implementation, which can be optimized using standard CNN operators:

![image](https://user-images.githubusercontent.com/8041160/78451637-9600bc00-763b-11ea-9faf-e50957d00525.png)

**Comment by zhanghang1989** [comment]
We have similar conclusion as a concurrent work that increasing number of groups/splits will increase the latency on GPU:

![image](https://user-images.githubusercontent.com/8041160/78451667-d19b8600-763b-11ea-8604-7ba892b8e19d.png)

Therefore, we didn't use large number of groups.

**Comment by GuideWsp** [comment]
Thanks for you relpy. Just from the code, it's difficult to judge the difference of the two method.
The standard SKNet has two groups, of course it can be set to other number.
But if the group number can be auto-optimize, it's a very good idea.
Hoping for the paper.

**Comment by zhanghang1989** [comment]
> Thanks for you relpy. Just from the code, it's difficult to judge the difference of the two method.
> The standard SKNet has two groups, of course it can be set to other number.
> But if the group number can be auto-optimize, it's a very good idea.
> Hoping for the paper.

ResNeSt introduce two additional concepts:
- cardinal groups (given by cardinality),
- radix (number of splits in each group, instead of the entire network),

which is shown more efficient than prior work.

**Comment by GuideWsp** [comment]
From the pictures, it has two difference between SKNet and ResNeSt.
1. Goups on Conv1x1;
2. Split Attention with Concat.

while in the code:
1. self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)
2. out = sum([att*split for (att, split) in zip(atten, splited)])

From the above code, I think the two are very similar.
I have not try the code(I will try it later), maybe something wrong about the code.

**Comment by zhanghang1989** [comment]
> From the pictures, it has two difference between SKNet and ResNeSt.
> 
> 1. Goups on Conv1x1 and Conv3x3 in Bottleneck;
> 2. Split Attention with Concat.
> 
> while in the code:
> 
> 1. self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)
> 2. out = sum([att*split for (att, split) in zip(atten, splited)])
> 
> From the above code, I think the two are very similar.
> I have not try the code(I will try it later), maybe something wrong about the code.

Please Refer to the figure for "Radix-major implementation".

The first figure is easy-to-understand for the reader, only for demonstration purpose. The "Radix-major implementation" shows the actual implementation. 

Anyway, I would suggest waiting for the arXiv, which should be next week.

**Comment by GuideWsp** [comment]
\

> > From the pictures, it has two difference between SKNet and ResNeSt.
> > 
> > 1. Goups on Conv1x1 and Conv3x3 in Bottleneck;
> > 2. Split Attention with Concat.
> > 
> > while in the code:
> > 
> > 1. self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)
> > 2. out = sum([att*split for (att, split) in zip(atten, splited)])
> > 
> > From the above code, I think the two are very similar.
> > I have not try the code(I will try it later), maybe something wrong about the code.
> 
> Please Refer to the figure for "Radix-major implementation".
> 
> The first figure is easy-to-understand for the reader, only for demonstration purpose. The "Radix-major implementation" shows the actual implementation.
> 
> Anyway, I would suggest waiting for the arXiv, which should be next week.

I agree, the figure is easy-to-understand. Hoping for the paper. 
Hoping for the arix, maybe it can give me more information.
Anyway, thanks and it's a interesting work.

**Comment by swjtulinxi** [comment]
> This work is inspired by pioneering work, and generalizes SENet/SKNet into featuremap groups and outperforms prior work.

hi，i have a question in the code ,why the groups=1??,

**Comment by GuideWsp** [comment]
Hey, ResNeSt can be treat as the combination of ResNeXt and SKNet, is that right?
will you provide a pretrained model with cardinal > 1?
if cardinal ==1, it can be treat as SKNet.

**Comment by zhanghang1989** [comment]
> Hey, ResNeSt can be treat as the combination of ResNeXt and SKNet, is that right?
> will you provide a pretrained model with cardinal > 1?
> if cardinal ==1, it can be treat as SKNet.

Increasing `radix` or `cardinality`

**Comment by zhanghang1989** [comment]
* Our work is inspired by SE-Net, SK-Net and ResNeXt. 
* SKNet is a multi-path network, which is not modularized and optimized for training/inference.
* ResNeSt generalize the channel-wise attention into group level features, which is more efficient.

**Comment by snowzjy** [comment]
In the file resnest/torch/resnet.py
line 41-66, the codes are more like SK-net block，not like resnest block. Is there something wrong?

        self.conv1 = nn.Conv2d(inplanes, group_width, kernel_size=1, bias=False)
        self.bn1 = norm_layer(group_width)
        self.dropblock_prob = dropblock_prob
        self.radix = radix
        self.avd = avd and (stride > 1 or is_first)
        self.avd_first = avd_first

        if self.avd:
            self.avd_layer = nn.AvgPool2d(3, stride, padding=1)
            stride = 1

        if dropblock_prob > 0.0:
            self.dropblock1 = DropBlock2D(dropblock_prob, 3)
            if radix == 1:
                self.dropblock2 = DropBlock2D(dropblock_prob, 3)
            self.dropblock3 = DropBlock2D(dropblock_prob, 3)

        if radix > 1:
            self.conv2 = SplAtConv2d(
                group_width, group_width, kernel_size=3,
                stride=stride, padding=dilation,
                dilation=dilation, groups=cardinality, bias=False,
                radix=radix, rectify=rectified_conv,
                rectify_avg=rectify_avg,
                norm_layer=norm_layer,
                dropblock_prob=dropblock_prob)

**Comment by zhanghang1989** [comment]
This is **equivalent** radix-major implementation for modularization acceleration. Please read the **paper appendix** for details.

The following 3 implementations are identical mathematically: 

![ResNeStV2](https://user-images.githubusercontent.com/8041160/79690015-24189d00-820d-11ea-835b-d12833ba2d81.png)
![ResNeStV22](https://user-images.githubusercontent.com/8041160/79690019-267af700-820d-11ea-8916-d0792695896d.png)
![ResNeStUV2](https://user-images.githubusercontent.com/8041160/79690021-28dd5100-820d-11ea-9930-8636d501cff2.png)

**Comment by GuideWsp** [comment]
![image](https://user-images.githubusercontent.com/9530892/79824927-1c522780-83ca-11ea-83cd-45c22680efb7.png)
We check out the splatConv2d, and found a little bug, and modify it. 
If possible, you can check it.

**Comment by zhanghang1989** [comment]
Please checkout our description of radix-major implementation.

**Comment by zhanghang1989** [comment]
Please note our ablation study models are available at https://github.com/zhanghang1989/ResNeSt/blob/master/ablation.md

**Comment by Xujianzhong** [comment]
@zhanghang1989 thanks!
